### PR TITLE
Adding dropout probability to `model_fn`.

### PR DESCRIPTION
**Pull request by znado**
- Pass `dropout_prob` and `aux_dropout_prob` to `model_fn(...)` everywhere, as per the rules. If no dropout is provided in the reference submissions, we default to the model-defined default.
- Trying to be consistent using `dropout_rate` as the name throughout the codebase.
- Numerous fixes to get reference_submission_test to pass for all workloads.
- Renaming `init_metrics_bundle` to `init_tokenizer` to be clearer.

Some careful things to check:
- the `update*_dropout` functions in `pytorch_utils.py` work as intended
- the defaults are consistently used across the code
- the librispeech input pipeline changes are correct (we need to pass tuples of (input, input_padding) and (target, target_padding) in order to conform to the API where the same three keys are allowed in `batch`)

### CODE REVIEW

**Review by runame — APPROVED** [review]
Thanks for the effort! I hope iterating through all modules each time a PyTorch `model_fn` is called won't introduce any noticeable overhead.

File: algorithmic_efficiency/pytorch_utils.py [hunk 1]
```diff
@@ -41,3 +41,22 @@ def logging_pass(*args):
...
     # initialize the process group
     dist.init_process_group('nccl')
+
+
+def update_dropout(model, dropout_rate):
+  # model.modules() returns the model itself as the first element.
+  for child in list(model.modules())[1:]:
+    if isinstance(child, torch.nn.Dropout):
+      child.p = dropout_rate
+    update_dropout(child, dropout_rate)
```
[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/pytorch_utils.py:51]: Aaaahhh ok, that explains a lot. Done.

File: algorithmic_efficiency/pytorch_utils.py [hunk 2]
```diff
@@ -41,3 +41,22 @@ def logging_pass(*args):
...
+
+
+def update_attention_dropout(model, attention_dropout_rate):
+  # model.modules() returns the model itself as the first element.
+  for child in list(model.modules())[1:]:
+    if isinstance(child, torch.nn.TransformerDecoderLayer):
```
[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/pytorch_utils.py:57]: nice catch, done.

File: algorithmic_efficiency/pytorch_utils.py [hunk 3]
```diff
@@ -41,3 +41,22 @@ def logging_pass(*args):
       logging.info = logging_pass
     # initialize the process group
     dist.init_process_group('nccl')
+
+
+def update_dropout(model, dropout_rate):
+  # model.modules() returns the model itself as the first element.
+  for child in list(model.modules())[1:]:
+    if isinstance(child, torch.nn.Dropout):
```
[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/pytorch_utils.py:49]: That's a good catch for the future, I checked and the current code doesn't have it. Added a note here to hopefully be seen in the future if we add more dropout.

File: algorithmic_efficiency/spec.py [hunk 4]
```diff
@@ -206,6 +206,8 @@ def model_fn(self,
                model_state: ModelAuxiliaryState,
                mode: ForwardPassMode,
                rng: RandomState,
+               dropout_rate: float,
```
[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/spec.py:209]: Good point, done.

File: algorithmic_efficiency/workloads/fastmri/fastmri_pytorch/workload.py [hunk 5]
```diff
@@ -273,6 +277,8 @@ def _eval_model_on_split(self,
         model_state,
         spec.ForwardPassMode.EVAL,
         model_rng,
+        dropout_rate=0.0,  # Not relevant for eval.
```
[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/workloads/fastmri/fastmri_pytorch/workload.py:280]: I think the issue is actually with jax, where if we pass in None then it will complain about subtracting a float and None inside flax.linen.Dropout

[the remark this answers is not in the document]
Comment (runame) [reply algorithmic_efficiency/workloads/fastmri/fastmri_pytorch/workload.py:280]: That makes sense -- `flax.linen.Dropout` actually checks if the dropout rate is 0 and then just returns the inputs. I guess we could leave it like it is in PyTorch for consistency, or just only change it to `None` for PyTorch.

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/models.py [hunk 6]
```diff
@@ -49,6 +49,7 @@ class ConformerConfig:
   time_mask_max_ratio: float = 0.05
   time_masks_per_frame: float = 0.0
   use_dynamic_time_mask_max_frames: bool = True
+  # DO NOT SUBMIT use or delete?
```
[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/models.py:52]: clarified with Sourabh and cleaned it up.

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/workload.py [hunk 7]
```diff
@@ -203,7 +213,11 @@ def eval_step_pmapped(
         params,
         batch,
         model_state,
-        spec.ForwardPassMode.EVAL)
+        spec.ForwardPassMode.EVAL,
+        rng,
+        dropout_rate=0.1,
```
[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/workload.py:218]: it isn't, I was using the defaults defined in the model, but I think you're right and it makes more sense to just pass 0.0.

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py [hunk 8]
```diff
@@ -21,12 +21,23 @@
...
+USE_PYTORCH_DDP, RANK, DEVICE, N_GPUS = pytorch_utils.pytorch_setup()
 
 MAX_INPUT_LENGTH = 320000
 
 
+def _update_model_dropout(model, residual_dropout_rate, input_dropout_rate):
+  # model.modules() returns the model itself as the first element.
+  for child in list(model.modules())[1:]:
+    if isinstance(child, Subsample):
+      child.input_dropout_rate = input_dropout_rate
```
[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:36]: ah, that makes sense. done!

File: algorithmic_efficiency/workloads/wmt/wmt_pytorch/models.py [hunk 9]
```diff
@@ -292,9 +303,12 @@ def forward(
...
 class PositionalEncoding(nn.Module):
 
-  def __init__(self, d_model: int, dropout: float = 0.1, max_len: int = 256):
+  def __init__(self,
+               d_model: int,
+               dropout_rate: float = 0.1,
+               max_len: int = 256):
     super().__init__()
-    self.dropout = nn.Dropout(p=dropout)
+    self.dropout_rate = nn.Dropout(p=dropout_rate)
```
[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/workloads/wmt/wmt_pytorch/models.py:311]: this was unintended, done.

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/workload.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (runame) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_jax/workload.py]: Yeah I guess this would look less confusing in the code.

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (runame) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py]: Yes, since this function is written specifically for a single model your solution should work (I didn't look at the model def before, hence my more general suggestion). Just to make sure, we do not want to set the dropout rate for `conformer_model.FeedForwardModule.dropout`?

[the remark this answers is not in the document]
Comment (znado) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py]: right, we just want to modify `feed_forward_residual_dropout_rate`. we should clean up the librispeech models/configs to bake in all the defaults (we won't be modifying them anyways)
